#01. Importing libraries

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

#02. Importing data

In [2]:
#create string path for main project folder
path = r'/Users/AngieUS/Desktop/Instacart Project'

#import original products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

#check df_prods
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [3]:
#import wrangled orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

#check df_ords
df_ords.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             1           1   2398795        1             2   
2             2           2    473747        1             3   
3             3           3   2254736        1             4   
4             4           4    431534        1             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     NaN  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

#03. Mixed data

In [4]:
#create a dataframe
df_test = pd.DataFrame()

In [5]:
#create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
#check df_test
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
#check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
#convert column 'mix' to string
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
#check df_test data types
df_test.dtypes

mix    object
dtype: object

#04. Missing data

In [12]:
#find missing data in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
#view missing data from product_name in df_prods using df_nan (new dataframe)
df_nan = df_prods[df_prods['product_name'].isnull() == True]

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
#display dimensions of df_prods - rows,columns
df_prods.shape

(49693, 5)

In [16]:
#create new dataframe df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

#display dimensions of df_prods_clean - rows, columns
df_prods_clean.shape

(49677, 5)

#05. Duplicate data

In [17]:
#look for duplicates in df_prods
df_dups = df_prods_clean[df_prods_clean.duplicated()]

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
#display dimensions of df_prods_clean - rows, columns
df_prods_clean.shape

(49677, 5)

In [19]:
#create dataframe without the duplicate rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

#display dimensions of df_prods_clean_no_dups - rows, columns
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
#export df_prods_clean_no_dups as "products_checked.csv"
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

#06. Task Questions

In [27]:
#display descriptive analysis of df_ords
df_ords.describe().round(2)

order_id     user_id  order_number  orders_day_of_week  \
count  3421083.00  3421083.00    3421083.00          3421083.00   
mean   1710542.00   102978.21         17.15                2.78   
std     987581.74    59533.72         17.73                2.05   
min          1.00        1.00          1.00                0.00   
25%     855271.50    51394.00          5.00                1.00   
50%    1710542.00   102689.00         11.00                3.00   
75%    2565812.50   154385.00         23.00                5.00   
max    3421083.00   206209.00        100.00                6.00   

       order_hour_of_day  days_since_prior_order  
count         3421083.00              3214874.00  
mean               13.45                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

#First, order_id, user_id & order_number are keys/labels so they can be disregarded as they aren't numeric for calculation. 
#Count seems to be consistent at 3421083 except for 'days_since_prior_order' is less 206209 at 3214874 - warrants further investigation.
#Mean & Standard Deviation appear normal.
#Min is all 0 values which is plausible for day 0, hour 0 and prior order 0. Max is also expected with highest/last day 6, hour 23, and month 30.
#25%, 50%, and 75% seem to be as expected.

In [28]:
#check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

No columns of mix types printed.

In [31]:
#find missing data in df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#The only column with missing data is 'days_since_prior_order' in the exact COUNT discrepancy: 206209.
#Blanks in this column could be due to not being entered or for orders with the same date the data was taken causing a conflict... or the date could be out of range (too far back). Futher investigation is needed to determine the probable cause.

In [32]:
#display dimensions of df_ords - rows,columns
df_ords.shape

(3421083, 6)

In [33]:
#create new dataframe df_ords_clean
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

#display dimensions of df_ords_clean - rows, columns
df_ords_clean.shape

(3214874, 6)

#I chose to delete the missing data for several reasons:
#1. I need to use the column data in my analysis so I need it cleaned up.
#2. Because the amount of records with missing data was only 6% of the total data, I concluded the most efficient method would be remove the records.
#3. As you can see from the descriptive analysis below, removing the records did not make an impact on the numbers.

In [34]:
#display descriptive analysis of df_ords_clean
df_ords_clean.describe().round(2)

order_id     user_id  order_number  orders_day_of_week  \
count  3214874.00  3214874.00    3214874.00          3214874.00   
mean   1710675.43   102970.08         18.19                2.78   
std     987546.57    59534.11         17.80                2.04   
min          1.00        1.00          2.00                0.00   
25%     855547.25    51385.00          6.00                1.00   
50%    1710834.50   102662.00         12.00                3.00   
75%    2565904.75   154372.00         25.00                5.00   
max    3421083.00   206209.00        100.00                6.00   

       order_hour_of_day  days_since_prior_order  
count         3214874.00              3214874.00  
mean               13.44                   11.11  
std                 4.23                    9.21  
min                 0.00                    0.00  
25%                10.00                    4.00  
50%                13.00                    7.00  
75%                16.00                   15.00  
max                23.00                   30.00

In [35]:
#look for duplicates in df_ords_clean
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

#No duplicate records found.

In [36]:
#export df_ords_clean as "orders_checked.csv"
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))